### ISSUE TITLE
Improve symbolic reasoning on intervals

### ISSUE THREAD

**Issue by willow-ahrens** [willow-ahrens/Finch.jl#37]
It would be great to define some way to simplify "c = choose(a, b)" when we need to lower c. We should choose the term with more information.

**Comment by willow-ahrens** [comment]
To elaborate, this problem usually arises during the calculation of loop bounds. When we expect the dimensions of the arrays to match, (for example, in `@loop i x[i] = y[i] + z[i]`, we end up with multiple sources for the range of i (`length(y)` or `length(z)` are both good) which we expect to be equivalent. It would be great if we could introduce a few rules to deterministically choose endpoints which are less expensive to calculate (literals are preferred over dynamic values, which we prefer over function calls).

Furthermore, the `Extent` type is a little complicated. It has fields for the endpoints of the range, but also fields for lower and upper bounds on the length of the range. We needed these bounds because we eagerly simplify endpoints where two things are equal, losing information. For example, the range of `i` might be calculated to be `length(y)`, when it would more accurately be described as `equiv(length(y), length(z))`, where `equiv` is a function which asserts its arguments are equal and returns any of them. If we represent ranges this way, we might be able to do better math with them. For example, if we wish to say that a range `a:b` has length 1, we could write it as `equiv(a, b-1):equiv(a + 1, b)`. If we wish to say that a range `a:b` fits inside a range `a:c`, we might write it as `a:equiv(b, min(b, c))`.

The goal for this issue is to add the `equiv` function and associated machinery to get this working and simplify extent handling.

**Comment by willow-ahrens** [comment]
There's a partial approach in https://github.com/willow-ahrens/Finch.jl/pull/150, but I'm beginning to realize that replacing the existing system of special cases with a general symbolic reasoning system is mostly sleight of hand. Ultimately, there are many different canonical forms for our system, each supporting certain kinds of queries and requiring certain amounts of combinatorial expansion to support. 

I can at least spell out the properties Finch wants to declare and the queries Finch wants to support. 

Properties of values that Finch declares about loop bounds:

1. The interval is nonempty
2. The interval has an exact length
3. The stopping endpoint is less than some other variables (with constant values) in scope.
4. The start or endpoint is equal to some other variables in scope.

Queries that Finch wants to answer about intervals:

1. Is an interval nonempty? (Is stop - start > 0?) (is min(stop - start, 0) == 0?)
2. Is min(x, y) equal to min(x)?

**Comment by wraith1995** [comment]
So, first, I think that we should note why 
```
@loop i x[i] = y[i] + z[i]
```
becomes a bounds inference issue. At the core, is the "directionality" i.e do output bounds determine inputs or inputs determine outputs. In Halide, this is a non-issue because outputs bounds determine input bounds. In Finch, this is not the case, and the bounds might be determined via the opposite procedure, so we end up with a choice when we must merge bounds. I think the first step to fix this is to think about the "directionality" because these solutions in some sense reverse that directionality at different point in the algorithm. 

Even if fixing (2) is not an issue (I don't think this is the case), I think we should fix this. I think what you are proposing here is that in certain circumstances (which ones though?), `=` changes the directionality and we could deal this via a few strategies, that I will list. But first, I will note that I think we should keep the forwards pass because I think it is worthwhile to consider that in some cases you might not want `=` to enforce equality in bounds between `y[i]` and `z[i]`

Okay, consider everything after this a DRAFT thought because I need to think about other things:

-Just go backwards - I think this would be too much of change/doesn't agree well with sparse.
- Going backwards in some places until we reach inputs so that we can add asserts. Then we can use the assumptions guaranteed by the asserts. For example, we could when we change direction in this circumstance, pick one bound and then make a backwards bound query to generate start program asserts that ensure the bounds have to be equal. These asserts then might be easier to simplify/might mostly be at the start of programs, making them cheaper.
- A system for accumulating assumptions into bounds info, which can hopefully normalize to simple single bounds. (I think this is your option (1) though I would do it with expressions directly as AtMost and Equiv should be expressive with unions, intersections, min, maxes, etc - which is probably a more complete system.)
- One could make an argument that doing (2) at the end after forwards inference might reduce just to running backwards inference on the forward pass. 

Some other thoughts:

- I think that AtMost/Equiv can be normalized to just mins, maxes, with added assumptions i.e I think we could represent some mixed direction algorithm with these and then convert when we want to use lots of Halide rules. 
- I think egraphs might be more useful than z3 - I think we should look at Caviar as they have lots of Halide's rewrite rules wirtten down and a good blue print to use egraphs in systems like this:  (https://github.com/caviar-trs/caviar and https://arxiv.org/pdf/2111.12116.pdf ).

**Comment by willow-ahrens** [comment]
@wraith1995 On the [extents2](https://github.com/willow-ahrens/Finch.jl/tree/wma/extents2) branch, I've added a sketch of what an e-graph in finch might look like. I had to do a few hacks to get things to work, but it seems to faithfully apply the rules and solves most of the (very simple) queries we currently give it. The idea is to move some of the current bounds logic over into the egraph ruleset. We should consider how much work it will take to harden whatever e-graph approach we use.

From my initial investigation, it will be difficult to make any guarantees about whether we can actually prove all of the queries we want to prove. For example, [here](https://github.com/JuliaSymbolics/Metatheory.jl/issues/154) is a simple ruleset which is clearly a subset of what we want to do, and will never saturate fully. So we would be relying on the rule scheduling heuristic of whatever e-graph package we are using to reliably find the kinds of simplifications we're interested in. The approach could probably be improved if every time we modify a bounds value (in ctx.bindings), we also run an e-graph simplifier on it, so that complexity doesn't build up.

In the [extents](https://github.com/willow-ahrens/Finch.jl/tree/wma/extents) branch, I moved all the bounds logic into expressions. In general, adding mins and maxes to all the expressions made things a little messy, so we would also have to think of a cleaner way to do that.
